rust/kernel/print.rs

```diff
@@ -190,6 +203,90 @@ macro_rules! print_macro (
+        static mut PRINTK_RS: core::mem::MaybeUninit<$crate::print::ratelimit_state> =
+            core::mem::MaybeUninit::<$crate::print::ratelimit_state>::uninit();
+        static mut PRINTK_RS_PTR: core::sync::atomic::AtomicPtr<$crate::print::ratelimit_state> =
+            core::sync::atomic::AtomicPtr::<$crate::print::ratelimit_state>::new(core::ptr::null_mut());
+        static mut PRINTK_RS_AVAILABLE: core::sync::atomic::AtomicBool =
+            core::sync::atomic::AtomicBool::new(false);
+
+        // SAFETY: [`PRINTK_RS_AVAILABLE`] and [`PRINTK_RS_PTR`] is used to
+        // control the intialization of [`PRINTK_RS`] - a local
+        // [`ratelimit_state`] initialized with [`DEFAULT_RATELIMIT_INTERVAL`]
+        // and [`DEFAULT_RATELIMIT_BURST`]. All of these three variables are
+        // static local variables to avoid dynamic memory allocation and no
+        // lock is used because printk can be used in any circumstance.
+        //
+        // The hidden macro [`call_printk`] should only be called by the
+        // documented printing macros which ensure the format string is one of
+        // the fixed ones.
+        // All `__LOG_PREFIX`s are null-terminated as they are generated
+        // by the `module!` proc macro or fixed values defined in a kernel
+        // crate.
+        unsafe {
+            if PRINTK_RS_AVAILABLE.compare_exchange(
+                false,
+                true,
+                core::sync::atomic::Ordering::SeqCst,
+                core::sync::atomic::Ordering::SeqCst
+            ).is_ok() {
+                // Zero out [`PRINTK_RS`] then initialize it using
+                // [`rust_helper_ratelimit_state_init`] helper function.
+                *PRINTK_RS.as_mut_ptr() = $crate::print::ratelimit_state::default();
+                $crate::print::rust_helper_ratelimit_state_init(PRINTK_RS.as_mut_ptr());
+
+                // Make PRINTK_RS_PTR point at PRINTK_RS so later we can call
+                // [`___ratelimit`] with it.
+                if PRINTK_RS_PTR.load(
+                    core::sync::atomic::Ordering::SeqCst
+                ).is_null() {
+                    PRINTK_RS_PTR.store(
+                        PRINTK_RS.as_mut_ptr(),
+                        core::sync::atomic::Ordering::SeqCst
+                    );
+                }
+            }
+
+            // [`PRINTK_RS`] can be uninitialized under concurrent situation,
+            // if so, we print message without ratelimit control.
+            // Otherwise we call __ratelimit to decide if
+            // enforce ratelimit.
+            if PRINTK_RS_PTR.load(core::sync::atomic::Ordering::SeqCst).is_null() ||
+                $crate::print::___ratelimit(
+                    PRINTK_RS_PTR.load(core::sync::atomic::Ordering::SeqCst),
+                    $crate::c_str!($where).as_char_ptr()
+                ) != 0 {
+                $crate::print::call_printk(
```

Comment:
I think a better approach might be providing an abstraction of the ratelimiter instead of having these macro expanded each time.

Reply:
There is a problem, we need a static local ratelimite_state at each occurrence. Can we do this without macros??

Reply:
You can have a `Ratelimiter` that wraps `ratelimit_state` and call a method on it each time.

Reply:
But how can we preserve the state??
Can you elaborate further?

Reply:
You'll still need a static variable, but just `static RATERLIMITER: Ratelimiter = Ratelimiter::new();` (or perhaps `Lazy<Ratelimiter>`).

Reply:
I might need to use Lazy.
Does it work well in multithreading??

Reply:
There are `SyncLazy` in std (but not core) and `once_cell::sync::Lazy`. We probably need to implement our own `Lazy` in kernel.